### PR TITLE
Don't iterate over peers in generic_proto::behaviour::poll

### PR DESCRIPTION
**Pull request by tomaka**
At the moment, every single call to `GenericProto::poll` leads to iterating every single peer we are connected to (including the ones for which we are connected only for Kademlia-related purposes).

This PR optimizes this a bit.

The only reason why we need to iterate is to query the various `Delay`s. After this PR, we would now duplicate all these `Delay`s and push these duplicates in a `FuturesUnordered`.

Polling a `FuturesUnordered` is `O(1)` over the number of elements that it contains because it tracks which child futures have woken up and which haven't.

By design, we never remove elements from `peers_check_after`, as the `FuturesUnordered` struct doesn't allow removing elements anyway. As such, these duplicate `Delay`s will trigger even when there's nothing to do. This should be negligible.

### CODE REVIEW

File: client/network/src/protocol/generic_proto/behaviour.rs [hunk 1]
```diff
@@ -627,12 +637,22 @@ impl GenericProto {
 
 		match mem::replace(occ_entry.get_mut(), PeerState::Poisoned) {
 			PeerState::Banned { ref until } if *until > now => {
+				let peer_id = occ_entry.key().clone();
 				debug!(target: "sub-libp2p", "PSM => Connect({:?}): Will start to connect at \
-					until {:?}", occ_entry.key(), until);
+					until {:?}", peer_id, until);
 				*occ_entry.into_mut() = PeerState::PendingRequest {
 					timer: futures_timer::Delay::new(*until - now),
```
Comment (romanb) [inline client/network/src/protocol/generic_proto/behaviour.rs:644]: Why do we need to keep this timer?

File: client/network/src/protocol/generic_proto/behaviour.rs [hunk 2]
```diff
@@ -649,13 +669,23 @@ impl GenericProto {
 				open,
 				banned_until: Some(ref banned)
 			} if *banned > now => {
+				let peer_id = occ_entry.key().clone();
 				debug!(target: "sub-libp2p", "PSM => Connect({:?}): But peer is banned until {:?}",
-					occ_entry.key(), banned);
+					peer_id, banned);
 				*occ_entry.into_mut() = PeerState::DisabledPendingEnable {
 					open,
 					timer: futures_timer::Delay::new(*banned - now),
```
Comment (romanb) [inline client/network/src/protocol/generic_proto/behaviour.rs:677]: Analogous question as in the other case.

File: client/network/src/protocol/generic_proto/behaviour.rs [hunk 3]
```diff
@@ -1363,7 +1393,14 @@ impl NetworkBehaviour for GenericProto {
...
+			let peer_state = match self.peers.get_mut(&peer_id) {
+				Some(s) => s,
+				// We intentionally never remove elements from `peers_check_after`, and it may
+				// thus contain peers which are now gone. This is a normal situation.
+				None => continue,
+			};
+
 			match peer_state {
 				PeerState::PendingRequest { timer, .. } => {
 					if let Poll::Pending = Pin::new(timer).poll(cx) {
```
Comment (romanb) [inline client/network/src/protocol/generic_proto/behaviour.rs:1406]: Related to the other comments, why do we need to keep this `timer` and poll it here?

File: client/network/src/protocol/generic_proto/behaviour.rs [hunk 4]
```diff
@@ -126,6 +126,15 @@ pub struct GenericProto {
...
 	peers: FnvHashMap<PeerId, PeerState>,
 
+	/// `peers` occasionally contains a few `Delay` objects that we would normally have to be
+	/// polled one by one. In order to avoid doing so, as an optimization, every `Delay` is
+	/// duplicated in `peers_check_after`. This stream yields `PeerId`s whose `Delay` is
+	/// potentially ready.
+	///
+	/// By design, we never remove elements from this list. Elements are removed only when the
+	/// `Delay` triggers. As such, this stream may produce obsolete elements.
+	peers_check_after: stream::FuturesUnordered<Pin<Box<dyn Future<Output = PeerId> + Send>>>,
```
Comment (romanb) [inline client/network/src/protocol/generic_proto/behaviour.rs:136]:
```suggestion
	delayed_dialing: stream::FuturesUnordered<Pin<Box<dyn Future<Output = PeerId> + Send>>>,
```

**Comment by romanb** [comment]
As mentioned in the comments, I don't yet understand why we need to duplicate the `Delay`s instead of just moving them.

**Comment by tomaka** [comment]
> As mentioned in the comments, I don't yet understand why we need to duplicate the Delays instead of just moving them.

When a peer is in the `PendingRequest` state, it might transition to `Enabled` as a response to a connection being established. When a peer is in the `Banned` state, it might transition to `Disabled` or `Incoming` as a response to a connection being established.

Since we can't remove elements from `FuturesUnordered`, we can't remove the associated `Delay` even when we know that it would no longer be needed. As such, we have a problematic situation: if a peer transitions from, let's say, `Banned` to `Disabled` to `Enabled` to `Banned`, and then the `Delay` of the first `Banned` triggers, we have no way know whether it concerns the first or the second `Banned`.

We could obviously add some sort of "delay id" system, but that complicates everything for not much gain.

> Why do we need to keep this timer?

The timers in the peers state act as the authority.

To see it differently: it's as if we implicitly cached the fact all the timers in the peers state were constantly not ready, and the timers in the `FuturesUnordered` are an indication as to when clear this cache.

**Comment by tomaka** [comment]
I'm getting a nice performance boost on my node.

Before this PR, with 100 peers, the network worker was consuming ~0.19 CPUs.
After this PR, with 86 peers (still waiting for more incoming peers), it is consuming ~0.125 CPUs.

**Comment by romanb** [comment]
> Since we can't remove elements from FuturesUnordered, we can't remove the associated Delay even when we know that it would no longer be needed. As such, we have a problematic situation: if a peer transitions from, let's say, Banned to Disabled to Enabled to Banned, and then the Delay of the first Banned triggers, we have no way know whether it concerns the first or the second Banned.

The `Banned` state has no `Delay` though, only `PendingRequest` and `DisabledPendingEnable`. These two states also still have the `timer_deadline` - I didn't suggest to remove that. So when a timer triggers, you could just check that the deadline has been reached. If not it is a no-op. The deadline is renewed whenever the state transitions to one of these, so old timers firing will just be no-ops. Wouldn't that work?

**Comment by romanb** [comment]
> We could obviously add some sort of "delay id" system, but that complicates everything for not much gain.

I think I would actually prefer that. That is, `PeerState::PendingRequest` and `PeerState::DisabledPendingEnable` just carry some timer ID and when a timer fires, it returns the peer ID and its timer ID, so when polling the timers we just need to check if the timer IDs match. I just find it very weird to maintain two identical `Delay`s where one is polled after the other already fired, but it may just be me.

**Comment by tomaka** [comment]
> I'm getting a nice performance boost on my node.

> Before this PR, with 100 peers, the network worker was consuming ~0.19 CPUs.
> After this PR, with 86 peers (still waiting for more incoming peers), it is consuming ~0.125 CPUs.

> (note that "Before this PR" is actually a build from Friday the 22nd, but I don't think anything else has been changed that could explain this difference)

My node's network worker consumption seems to stabilize around ~0.15 CPUs after this PR.
Still a good improvement.

**Pull request merged**